File: src/geopipe_agent/models/pipeline.py
"""Pipeline and step definition models."""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class StepDefinition:
    """Definition of a single pipeline step from YAML."""

    id: str
    use: str
    params: dict = field(default_factory=dict)
    when: str | None = None
    on_error: str = "fail"  # fail / skip / retry
    backend: str | None = None

    def __post_init__(self):
        if self.params is None:
            self.params = {}


@dataclass
class PipelineDefinition:
    """Parsed pipeline definition from YAML."""

    name: str
    steps: list[StepDefinition]
    description: str = ""
    crs: str | None = None
    variables: dict = field(default_factory=dict)
    outputs: dict = field(default_factory=dict)

File: web/frontend/src/types/pipeline.ts
/** Pipeline-level metadata */
export interface PipelineMeta {
  name: string
  description?: string
  crs?: string
  variables: Record<string, any>
  outputs: Record<string, string>
}

/** Step node data for Vue Flow */
export interface StepNodeData {
  use: string
  label: string
  params: Record<string, any>
  when?: string
  onError?: string
  backend?: string
  status?: 'idle' | 'running' | 'success' | 'error'
  errorMessage?: string
}

/** Step parameter schema from backend */
export interface StepParamSchema {
  type: string
  required: boolean
  description: string
  default?: any
  enum?: string[]
}

/** Step schema from backend */
export interface StepSchema {
  name: string
  category: string
  description: string
  params: Record<string, StepParamSchema>
  outputs: Record<string, { type: string; description?: string }>
  supports_backend: boolean
  backends?: string[]
  examples?: Array<{ description: string; params: Record<string, any> }>
}

/** Category group */
export interface StepCategory {
  name: string
  label: string
  steps: StepSchema[]
}

/** Pipeline info (saved pipeline) */
export interface PipelineInfo {
  id: string
  name: string
  version: number
  created_at: string
  updated_at: string
}

/** SSE event types */
export type PipelineEventType = 'progress' | 'done' | 'error'

export interface PipelineProgressEvent {
  step_id: string
  status: 'running' | 'success' | 'error'
  message: string
}

export interface PipelineDoneEvent {
  result: Record<string, any>
  total_time: number
}

export interface PipelineErrorEvent {
  code: string
  message: string
  step_id?: string
}
